src/compile.c

```diff
@@ -0,0 +1,402 @@
+
+#include "c2mir/c2mir.h"
+#include "cache.h"
+#include "compile.h"
+#include "decode.h"
+#include "mir-gen.h"
+#include "mir.h"
+#include "riscv_private.h"
+#include "utils.h"
+
+typedef struct {
+    char *code;
+    size_t code_size;
+    size_t curr;
+} jit_item_t;
+
+typedef struct {
+    MIR_context_t ctx;
+    struct c2mir_options *options;
+    uint8_t debug_level;
+    uint8_t optimize_level;
+} riscv_jit_t;
+
+typedef struct {
+    char *name;
+    void *func;
+} func_obj_t;
+
+#define SET_SIZE_BITS 10
+#define SET_SIZE 1 << SET_SIZE_BITS
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - SET_SIZE_BITS)) & ((SET_SIZE) -1)
+
+typedef struct {
+    uint32_t table[SET_SIZE][32];
+} set_t;
+
+static inline void set_reset(set_t *set)
+{
+    memset(set, 0, sizeof(set_t));
+}
+
+static bool set_add(set_t *set, uint32_t key)
+{
+    uint32_t index = HASH(key);
+    uint8_t count = 0;
+    while (set->table[index][count]) {
+        if (set->table[index][count++] == key)
+            return false;
+    }
+
+    set->table[index][count] = key;
+    return true;
+}
+
+static bool set_has(set_t *set, uint32_t key)
+{
+    uint32_t index = HASH(key);
+    uint8_t count = 0;
+    while (set->table[index][count]) {
+        if (set->table[index][count++] == key)
+            return true;
+    }
+    return false;
+}
+
+static bool insn_is_branch(uint8_t opcode)
+{
+    switch (opcode) {
+#define _(inst, can_branch) IIF(can_branch)(case rv_insn_##inst:, )
+        RISCV_INSN_LIST
+#undef _
+        return true;
+    }
+    return false;
+}
+
+typedef void (*gen_func_t)(riscv_t *, const rv_insn_t *, char *, uint32_t);
+static gen_func_t dispatch_table[N_RV_INSN];
+static char funcbuf[128] = {0};
+#define GEN(...)                   \
+    sprintf(funcbuf, __VA_ARGS__); \
+    strcat(gencode, funcbuf);
+#define UPDATE_PC(inc) GEN("  rv->PC += %d;\n", inc)
+#define NEXT_INSN(target) GEN("  goto insn_%x;\n", target)
+#define RVOP(inst, code)                                            \
+    static void gen_##inst(riscv_t *rv UNUSED, const rv_insn_t *ir, \
+                           char *gencode, uint32_t pc)              \
+    {                                                               \
+        GEN("insn_%x:\n"                                            \
+            "rv->X[0] = 0;\n"                                       \
+            "rv->csr_cycle++;\n",                                   \
+            (pc));                                                  \
+        code;                                                       \
+        if (!insn_is_branch(ir->opcode)) {                          \
+            GEN("  rv->PC += ir->insn_len;\n");                     \
+            GEN("  ir = ir + 1;\n");                                \
+            NEXT_INSN(pc + ir->insn_len);                           \
+        }                                                           \
+    }
+
+#include "jit_template.c"
+
+RVOP(jal, {
```

Comment:
You have to explain why we have to provide the customized handlers for branch related instructions.